src/ripple/app/tx/impl/CashCheck.h

```diff
@@ -34,7 +34,7 @@ class CashCheck
     }
 
     static
-    NotTEC
+    std::pair<NotTEC, TxConsequences>
```

Comment:
FWIW, `TxConsequences` is not just about blocking.  It's also about spending.  And now it also carries information about `Sequence` jumps.

What you're proposing could be done.  But I think the cost would be an extra `Transactor` entry point that returned `TxConsequences`.  That entry point would _always_ be called just after `preflight`.

Personally I like forcing the author of each `preflight` method to think about what their consequences are.  If `consequences()` were a separate method with a defaulted behavior, then it becomes easy for a Transactor author to not think about it and get the default behavior, which may not be appropriate.  But that's just my approach, and I acknowledge that there are other approaches.

Reply:
I might look at coding something up, to see how hard it would be (if I get some time, I'm behind on some other stuff right now). I do know it's more than just about blocking, but the added logic in `preflight` at least looked like it would be better centralized in `applySteps` rather than spreading it around. I'll try to take a closer look, it think it would be a win if we could do that.

Reply:
@scottschurr 
Of our 20 transactors:
* 12 of them create `TxConsequences` as `TxConsequences{ctx.tx}`
* 3 of them create `TxConsequences` as `TxConsequences{ctx.tx,TxConsequences::blocker};`
* 5 of them create them do something else.

After the consequences are created, all the transactors just return the object.

The problem I have with the current solution is it modifies every `preclaim` and changes the return type of `preclaim`, it also changes `preclaim` the same way in every transactor, and in 15 of 20 cases creates the `TxConsequences` object in the same way. Here's two ideas that try to fix this:

1) Every transaction has to define an `static constexpr ConsequencesFactory` that has one of three values: `Normal`, `Blocking`, and `Custom`. When it's `Custom`, then the transactor has to define a `MakeTxConsequences` static function. With this solution, the existing `preflight` functions do not require any modification. If a transactor doesn't define this variable, a compile error happens. This forces transactor authors to think about `TxConsequences`.

2) As you suggest, we can define an extra `Transactor` entry point. We could have a default one in `Transactor` that handles non-blocking consequences, and a transactor would define their own if they wanted to override this. (we could also not provide a default, if we wanted a compile error).

Branch with patch 1 (enum): https://github.com/seelabs/rippled/commits/consequences-enum
Branch with patch 2 (entry point with default): https://github.com/seelabs/rippled/tree/consequences-static-funs
Of these, I would rank them as:
1) Enum per transactor
2) `MakeTxConsequences` entry point _without_ a default in the base class (no patch provided, but it's easy to see how this would work from patch 2)
3) `MakeTxConsequences` entry point with a default.
4) Modify all the `preclaim` functions

Please note that I just coded these patches up to give an idea what the solution would look like. I'd want to review them carefully myself if we do decide to cherry pick one of them.

Reply:
The `enum` approach is clever and not overly subtle.  Nice use of `if constexpr` which gives a bunch of compile-time validation.  Sure, I'd be fine switching to that approach.  Thanks for coding it up. :+1:

Reply:
I have the enum approach integrated.  And it _almost_ works.

It looks like an old Visual C++ compiler is not good enough at `if constexpr` to support that approach.  The three Travis builds that use MSVC 19.16.27035.0 are failing with messages like this (see https://travis-ci.com/github/scottschurr/rippled/jobs/305215758):
```
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(56): error C2039: 'makeTxConsequences': is not a member of 'ripple::DeleteAccount'
..\src\ripple/app/tx/impl/DeleteAccount.h(29): note: see declaration of 'ripple::DeleteAccount'
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(73): note: see reference to function template instantiation 'std::pair<ripple::NotTEC,ripple::TxConsequences> ripple::invoke_preflight_helper<ripple::DeleteAccount>(const ripple::PreflightContext &)' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\include\chrono(861): note: see reference to class template instantiation 'std::chrono::duration<__int64,std::ratio<1,1>>' being compiled
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(56): error C3861: 'makeTxConsequences': identifier not found
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(56): error C2039: 'makeTxConsequences': is not a member of 'ripple::CancelCheck'
..\src\ripple/app/tx/impl/CancelCheck.h(27): note: see declaration of 'ripple::CancelCheck'
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(77): note: see reference to function template instantiation 'std::pair<ripple::NotTEC,ripple::TxConsequences> ripple::invoke_preflight_helper<ripple::CancelCheck>(const ripple::PreflightContext &)' being compiled
```
All of the other builds, including the build using VC2019, are working.  So Microsoft knows they have a bug and they've fixed it.  But, at least for now, it looks like that technique is not gonna be useable by us.

I may be able to get the same effect (in a way that all of our compilers support) by specializing templates.  That's what I'm gonna try next.

Bummer.